rosidl_generator_type_hash/rosidl_generator_type_hash/__init__.py

```diff
@@ -0,0 +1,320 @@
+    package_name = args['package_name']
+    output_dir = Path(args['output_dir'])
+    idl_tuples = args['idl_tuples']
+    include_paths = args.get('include_paths', [])
+
+    # Lookup for directory containing dependency .in.json files
+    include_map = {
+        package_name: output_dir
+    }
+    for include_tuple in include_paths:
+        include_parts = include_tuple.rsplit(':', 1)
+        assert len(include_parts) == 2
+        include_package_name, include_base_path = include_parts
+        include_map[include_package_name] = Path(include_base_path)
+
+    generated_files = []
+    hashers = []
+
+    # First generate all .in.json files (so referenced types can be used in expansion)
+    for idl_tuple in idl_tuples:
+        idl_parts = idl_tuple.rsplit(':', 1)
+        assert len(idl_parts) == 2
+        locator = definition.IdlLocator(*idl_parts)
+        try:
+            idl_file = parse_idl_file(locator)
+        except Exception as e:
+            print('Error processing idl file: ' +
+                  str(locator.get_absolute_path()), file=sys.stderr)
+            raise(e)
+
+        idl_rel_path = Path(idl_parts[1])
+        generate_to_dir = (output_dir / idl_rel_path).parent
+        generate_to_dir.mkdir(parents=True, exist_ok=True)
+
+        hasher = InterfaceHasher.from_idl(idl_file)
+        generated_files.extend(
+            hasher.write_json_in(output_dir))
+        hashers.append(hasher)
+
+    # Expand .in.json and generate .sha256.json hash files
+    for hasher in hashers:
+        generated_files.extend(
+            hasher.write_json_out(output_dir, include_map))
+        generated_files.extend(
+            hasher.write_hash(output_dir))
+
+    return generated_files
+
+
+# This mapping must match the constants defined in type_description_interfaces/msgs/FieldType.msg
+# NOTE: Nonexplicit integer types are not defined in FieldType (short, long, long long).
+# If a ROS IDL uses these, this generator will throw a KeyError.
+FIELD_TYPES = {
+    # 0 reserved for "Not set"
+    'nested_type': 1,
+    'int8': 2,
+    'uint8': 3,
+    'int16': 4,
+    'uint16': 5,
+    'int32': 6,
+    'uint32': 7,
+    'int64': 8,
+    'uint64': 9,
+    'float': 10,
+    'double': 11,
+    'long double': 12,
+    'char': 13,
+    'wchar': 14,
+    'boolean': 15,
+    'octet': 16,  # byte
+    definition.UnboundedString: 17,
+    definition.UnboundedWString: 18,
+    # NOTE: rosidl_parser does not define fixed string types
+    # FIXED_STRING: 19
+    # FIXED_WSTRING: 20
+    definition.BoundedString: 21,
+    definition.BoundedWString: 22,
+}
+
+FIELD_TYPE_BLOCK_SIZE = 48
+NESTED_FIELD_TYPE_OFFSETS = {
+    definition.Array: FIELD_TYPE_BLOCK_SIZE,
+    definition.BoundedSequence: FIELD_TYPE_BLOCK_SIZE * 2,
+    definition.UnboundedSequence: FIELD_TYPE_BLOCK_SIZE * 3,
+}
+
+
+def serialize_field_type(ftype: definition.AbstractType) -> dict:
+    result = {
+        'type_id': 0,
+        'length': 0,
+        'string_length': 0,
```

Comment:
string length never gets set?

Reply:
Good catch, I wasn't handling sequences/strings correctly. I fixed and added a simple test for it. Wondering if I should hand-compute some full sample message JSONs and compare to this generator's hash of the same `rosidl_parser.definition` object. Always the opportunity for human error but at least it'd be a smoke check

Reply:
could commit the definition,  message JSON, hash to the repo for a trivial message definition?